doc/design/docker-cp.md

```diff
@@ -0,0 +1,144 @@
+ 
+ This is the case where guest tools is not needed since the guest in off. This means that all action will need to occur on the VCH. There are several concerns and many of them are listed in this ticket [here](https://github.com/vmware/vic/issues/4105)
+ 
+ The most concern surrounding this piece of work will be surrounding VCH restart and confirming that the disk is still left in a good state if a copy was occuring mid restart. Additionally we need to block a container from starting when we have the r/w layer mounted. While that container is blocked from being started we might look at the possibility of returning a 500 error for a start invocation on a blocked container. We must also block other copy operations until the current one is successful, one major reason is because we cannot do any kind of sanity checking on available space with concurrent writes to the same disk. 
+ 
+ Additionally, it must be ensured that regardless of error or success the disk must be detached from the VCH. We must ensure that this also occurs when we boot up the VCH in the event that the disk is attached to the VCH and is hard shutdown. If the VCH does not already have a disk detach functionality at boot time we should be adding this. This is an ideal time to invoke calls to `fsck` and `journal` in order to recover from mid copy shutdowns. Since the VCH is meant to be stateless(and the portlayer) we should aim to jettison and disk that is attached to the VCH at boot time if we do write this portion of the portlayer(We should even make it a separate ticket as testing and ensuring quality on this could take time.)
+ 
+ In An Ideal scenario without any failures the process will be as such : 
+ 
+ ```
+ 1) get request from personality for a file copy, it will have the stream to readfrom/write to as well as a path and a container ID.
+ 
+ 2) depending on the endpoint hit depends on the to/from scenario either way in an *off* scenario we would first get the container object and confirm that it is off. 
+ 
+ 3) Once confirmed that the container is off we will need to immediately block starts(and even wait if one is triggered between the status check and requesting the lockdown). In addition to this, we must block other `cp` operations until the current one is complete. 
+ 
+ 4) Then we will mount the r/w layer at a temp directory location. and assemble that path along with the target copy path.
+ 
+ 5) Begin streaming the tar file to the mounted disk. 
+ 
+ 6) We will need to untar this in place while checking for existing files and behaving according to the overwrite dir flag. We will need this to be in place since we could be operating with limited space, we want to do this in a fashion that we have as little storage overhead during unpacking as possible. Checking flags for `untar` may help significantly with this. Otherwise we may need to read the tar headers themself before attempting the write and determining if we have the size to do this. 
```

Comment:
If we do concurrent writes then checking ahead of time may not be very useful. I do not see any mention of this in their documentation so I will take a look in their code. I also suspect it is fail hard on write error(OOS error). That would be more conducive to multiple concurrent writes.